TYPING: Added types for some tests

TomAugspurger opened the pull request:
Working around a strange typing issue. See
https://github.com/pandas-dev/pandas/pull/28394#issuecomment-5456527290
for more, but the types on these were being inferred incorrectly by
mypy with just the addition of the `allows_duplicate_labels` kwarg.

cc @simonjayhawkins & @WillAyd. Hopefully I got things correct. I copied the use of TypeVar for FrameOrSeries with IndexOrSeries.

TomAugspurger committed:
TYPING: Added types for tests files

Working around a strange typing issue. See
https://github.com/pandas-dev/pandas/pull/28394#issuecomment-5456527290
for more, but the types on these were being inferred incorrectly by
mypy with just the addition of the `allows_duplicate_labels` kwarg.

jreback commented on pandas/_typing.py:
```diff
@@ -32,6 +33,7 @@
 FilePathOrBuffer = Union[str, Path, IO[AnyStr]]
 
 FrameOrSeries = TypeVar("FrameOrSeries", bound="NDFrame")
+IndexOrSeries = TypeVar("IndexOrSeries", bound="IndexOpsMixin")
```
is this meant. it to include array likes, eg EA?

TomAugspurger replied:
Nope, just Index and Series.

jreback commented on pandas/tests/arithmetic/test_numeric.py:
```diff
@@ -74,33 +75,22 @@ def test_compare_invalid(self):
+    pd.Index,
+]  # type: List[Union[Type[pd.Index], Type[pd.RangeIndex], Type[pd.Series]]]
+left = [pd.RangeIndex(10, 40, 10)]  # type: List[Union[Index, Series]]
+for cls in index_or_series_params:
```
isn’t this any_numeric_dtype ?

TomAugspurger replied:
This seems to include float16. Aside from that, they look the same at a glance.

jreback replied:
we barely support float16 would just remove and use the fixture

WillAyd commented:
Any chance you can post the failures this was solving? Would be helpful as background

simonjayhawkins commented:
> Any chance you can post the failures this was solving? Would be helpful as background

```
pandas\tests\test_strings.py:206: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:240: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:379: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:391: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:431: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:444: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:498: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:600: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:660: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_strings.py:661: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_base.py:519: error: List item 0 has incompatible type "Type[Index]"; expected "Type[PandasObject]"
pandas\tests\test_base.py:550: error: List item 0 has incompatible type "Type[Index]"; expected "Type[PandasObject]"
pandas\tests\test_base.py:615: error: List item 0 has incompatible type "Type[Index]"; expected "Type[PandasObject]"
pandas\tests\test_base.py:1093: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_base.py:1123: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_base.py:1147: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_base.py:1335: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\test_base.py:1399: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\io\json\test_json_table_schema.py:434: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\io\json\test_json_table_schema.py:441: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\indexing\test_coercion.py:518: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\indexing\test_coercion.py:544: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\indexing\test_coercion.py:566: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\indexing\test_coercion.py:786: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\indexing\test_coercion.py:798: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\indexing\test_coercion.py:836: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\indexing\test_coercion.py:868: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\dtypes\test_concat.py:43: error: List item 0 has incompatible type "Type[Index]"; expected "Type[PandasObject]"
pandas\tests\arithmetic\test_numeric.py:87: error: List comprehension has incompatible type List[PandasObject]; expected List[RangeIndex]
pandas\tests\arithmetic\test_numeric.py:101: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\arithmetic\test_numeric.py:126: error: List comprehension has incompatible type List[PandasObject]; expected List[RangeIndex]
pandas\tests\arithmetic\test_numeric.py:140: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
pandas\tests\arrays\test_array.py:275: error: List item 0 has incompatible type "Type[Series]"; expected "Type[PandasObject]"
```

WillAyd commented on pandas/tests/arithmetic/test_numeric.py:
```diff
@@ -74,33 +75,22 @@ def test_compare_invalid(self):
     @pytest.mark.parametrize(
-        "left",
-        [pd.RangeIndex(10, 40, 10)]
-        + [
```
I imagine you were getting an error when adding lists together. This is an unfortunate limitation of mypy that you cannot use the `+` operator and get a Union of List types.

If you wanted to minimize diff you could alternately just unpack the second argument, so `[[pd.RangeIndex(10, 40 10)], *[cls([10, 20, 30], dtype=dtype) for dtype in [...`

mypy history goes back to:

https://github.com/python/typeshed/pull/2653

Definitely unfortunate

TomAugspurger replied:
Yep, that's exactly what I hit.

WillAyd commented on pandas/conftest.py:
```diff
@@ -790,6 +791,17 @@ def tick_classes(request):
+index_or_series_params = [pd.Index, pd.Series]  # type: IndexOrSeries
+
+
+@pytest.fixture(params=index_or_series_params, ids=["series", "index"])
```
Was this inferring as `List[PandasObject]` when used directly as an argument to `params`?

TomAugspurger replied:
Yes, unfortunately.

WillAyd commented on pandas/tests/test_base.py:
```diff
@@ -35,10 +36,12 @@
 from pandas.core.indexes.datetimelike import DatetimeIndexOpsMixin
 import pandas.util.testing as tm
 
+index_or_series = [Index, Series]  # type: List[Type[IndexOpsMixin]]
```
Are you not reusing the definition in `pandas._typing` here for a particular reason?

TomAugspurger replied:
Missed this one.

WillAyd reviewed (changes requested):
Looks pretty good. Some minor things I think can help the typing and legibility.

Nice to parametrize here as well

TomAugspurger committed:
start test

WillAyd commented on pandas/conftest.py:
```diff
@@ -790,6 +791,17 @@ def tick_classes(request):
     return request.param
 
 
+index_or_series_params = [pd.Index, pd.Series]  # type: IndexOrSeries
```
Should this be `List[Type[IndexOrSeries]]`? Surprised this passed checks as is; seems like something wonky going on

simonjayhawkins replied:
good catch.

agree that strange it doesn't fail

TomAugspurger commented on pandas/tests/test_strings.py:
```diff
@@ -8,10 +9,13 @@
 import pandas.util.testing as tm
 from pandas.util.testing import assert_index_equal, assert_series_equal
 
+index_or_series_params = [Index, Series]  # type: List[Type[PandasObject]]
```
With this, I get

```
pandas/tests/test_strings.py:18: error: Type variable "pandas._typing.IndexOrSeries" is unbound
```

Any suggestions?

simonjayhawkins replied:
probably don't need to create IndexOrSeries in _typing. so the union or the shared class is fine, but should probably be consistent.

WillAyd replied:
@simonjayhawkins do you know generally how mypy is inferencing these? I guess from the original list of errors I'm not sure where `Type[PandasObject]` even comes into play. Wondering if there isn't a core issue we are missing here, but I may also just be overlooking

simonjayhawkins replied:
my first reaction is that it is a mypy error. `Revealed type is 'builtins.list[def (data: Any =, Any =, Any =, name: Any =, Any =, Any =, **Any) -> pandas.core.base.PandasObject]'` seems iffy.

but agree, that it may be better to try and determine the cause rather than making these changes.

TomAugspurger replied:
That `reveal_type` trick is handy. It'd be nice if there was a way to print debug info about a particular chain of inference: all the decisions that led to mypy inferring that type.

WillAyd replied:
I wonder if types are getting mangled from pytest

simonjayhawkins replied:
i don't think pytest is relevant...

```
from typing import List, Type, cast

import pandas as pd

reveal_type([pd.Index, pd.Series])

[pd.Index, pd.Series]

cast(List[Type[pd.core.base.PandasObject]], [pd.Index, pd.Series])
```

```
$ mypy test.py
test.py:5: note: Revealed type is 'builtins.list[def (data: Any =, Any =, Any =, name: Any =, Any =, Any =, Any =) -> pandas.core.base.PandasObject]'
test.py:5: error: List item 0 has incompatible type "Type[Index]"; expected "Type[PandasObject]"
test.py:7: error: List item 0 has incompatible type "Type[Index]"; expected "Type[PandasObject]"
(pandas-dev) 
```

simonjayhawkins replied:
i don't think the `reveal_type([pd.Index, pd.Series])` should be giving an error which is why i suspect a mypy issue.

WillAyd replied:
Hmmm.....so when I run the above this is all I get:

```sh
$ mypy -V
mypy 0.720
$ mypy test.py
test.py:5: note: Revealed type is 'builtins.list[builtins.type*]'
```

Wonder why we get different output.

simonjayhawkins replied:
need to clone #28394

TomAugspurger commented:
Here's a somewhat reduced example

```python
# ok.py
from typing import List, Type, cast


class IndexOpsMixin:
    pass


class PandasObject:
    pass


class Series(IndexOpsMixin, PandasObject):
    def __init__(
        self, data=None, index=None, dtype=None, name=None, copy=False, fastpath=False,
    ):
        pass


class Index(IndexOpsMixin, PandasObject):
    def __new__(
        cls,
        data=None,
        dtype=None,
        copy=False,
        name=None,
        fastpath=None,
        tupleize_cols=True,
        **kwargs
    ) -> "Index":
        result = object.__new__(cls)
        return result



reveal_type([Index, Series])

[Index, Series]

cast(List[Type[PandasObject]], [Index, Series])
```

and

```python
# bug.py
from typing import List, Type, cast


class IndexOpsMixin:
    pass


class PandasObject:
    pass


class Series(IndexOpsMixin, PandasObject):
    def __init__(
        self, data=None, index=None, dtype=None, name=None, copy=False, fastpath=False, allow_duplicate_labels=False,
    ):
        pass


class Index(IndexOpsMixin, PandasObject):
    def __new__(
        cls,
        data=None,
        dtype=None,
        copy=False,
        name=None,
        fastpath=None,
        tupleize_cols=True,
        **kwargs
    ) -> "Index":
        result = object.__new__(cls)
        return result



reveal_type([Index, Series])

[Index, Series]

cast(List[Type[PandasObject]], [Index, Series])
```

so the only difference is the new keyword.


The outputs are

```python
bug.py:35: note: Revealed type is 'builtins.list[def (data: Any =, Any =, Any =, name: Any =, Any =, Any =, Any =) -> bug.IndexOpsMixin]'
bug.py:35: error: List item 0 has incompatible type "Type[Index]"; expected "Type[IndexOpsMixin]"
bug.py:37: error: List item 0 has incompatible type "Type[Index]"; expected "Type[IndexOpsMixin]"
ok.py:35: note: Revealed type is 'builtins.list[builtins.type*]'
```

TomAugspurger commented:
A couple observations:

1. This goes away when there are the same number of kwargs (adding an unused argument to Index solves things).
2. Removing **kwargs from `Index.__new__` solves things.

TomAugspurger commented:
May be onto something... This passes `mypy --strict` without any errors. I changed

* added a type to kwargs
* added a `type: Index` to the result of `Index.__new__`

```
# bug.py
from typing import List, Type, cast, Any, Mapping, Hashable


class Base:
    pass


class Series(Base):
    def __init__(self, a: Any = None, z: bool=False) -> None:
        pass


class Index(Base):
    def __new__(cls, a: Any = None, **kwargs: Mapping[Hashable, Any]) -> "Index":
        result = object.__new__(cls)  # type: Index
        return result


reveal_type([Index, Series])
[Index, Series]
cast(List[Type[Base]], [Index, Series])

```

WillAyd commented:
I asked about this on the typing Gitter and one of the devs shared this as where the "best" super type determination gets made:

https://github.com/python/mypy/blob/09c1fc7a4c19fe76c2ca9f0c58eaeccfe1ef9ed5/mypy/join.py#L340

Obviously here we have PandasObject and IndexOpsMixin as shared super types, and in this case it appears mypy infers PandasObject

Haven't stepped through the code but may provide some insight. In an case from comments seems like this behavior might be tenuous at best, so something to consider

simonjayhawkins commented:
might be clearer to use `return cast("Index", result)` instead of `  # type: Index`

TomAugspurger commented:
Unfortunately, I can't get the suggestion in https://github.com/pandas-dev/pandas/pull/29205#issuecomment-546038272 working on my branch.

TomAugspurger commented:
Given that this *seems* like a mypy bug, any objection to just adding `# type: ignore` to the lines causing issues in #28394? Hopefully that's OK since they're just test files...

simonjayhawkins commented:
my preference would be a `# type: ignore` to changing things if it maybe a mypy issue. 

`--warn-unused-ignores` can be used to remove if issues are fixed in mypy releases, so I don't think difficult to track.

WillAyd commented:
> Given that this _seems_ like a mypy bug,

I don't think this is a bug. It's just ambiguous as to what our intent is here.

Does making the TypeVar point to PandasObject work? I guess having a TypeVar use a Mixin might not make a ton of sense anyway

simonjayhawkins commented:
another alternative to `# type: ignore` could be to simply replace `[Index, Series]` with `cast(Any, [Index, Series])` as an interim measure.

simonjayhawkins commented:
> I don't think this is a bug.

```
from typing import List, Type, cast


class Base:
    pass


class Series(Base):
    def __init__(self, a=None, z=None):
        pass


class Index(Base):
    def __new__(cls, a=None, **kwargs) -> "Index":
        result = object.__new__(cls)
        return result


reveal_type([Index, Series])

cast(List[Type[Base]], [Index, Series])

b: List[Type[Base]] = [Index, Series]
reveal_type(b)
```
```
$ mypy bug.py
bug.py:19: note: Revealed type is 'builtins.list[def (a: Any =, Any =) -> bug.Base]'
bug.py:19: error: List item 0 has incompatible type "Type[Index]"; expected "Type[Base]"
bug.py:24: note: Revealed type is 'builtins.list[Type[bug.Base]]'
(pandas-dev) 
```

The assignment to `b` does not fail. Is this not the type that mypy should be inferring?

WillAyd commented:
That's a great example. I think the expression will always be ambiguous as to the intent whereas in the assignment we can clearly state what is expected. Can't we do that here in this PR but change `List[Type[IndexOrSeries]]` to just be `List[Type[PandasObject]]` then? I don't think it makes sense to use a Mixin here anyway, especially since these are really being used as constructors

A Literal `Type[Series]` or `Type[Index]`` would probably be best, but we don't have that support yet. If the `Type[PandasObject]` doesn't work then sure can ignore; I guess I'm just hesitant to do that because it's not clear what would need to happen to ever un-ignore

simonjayhawkins commented:
> I guess I'm just hesitant to do that because it's not clear what would need to happen to ever un-ignore

to be clear. mypy is infering a type for a list of two items, and then is immediately producing an error with the inferred type.

i.e. `reveal_type([Index, Series])` produces a error

the message in the error includes `expected "Type[Base]"` and yet explicity specifying the type as this type does not produce an error. i.e. `b: List[Type[Base]] = [Index, Series]`

so pretty sure this is a bug and what would need to happen to un-ignore is the bug to be fixed.

TomAugspurger commented:
I think I'll close this. Will ignore as needed on my other PR.
